bpf: fix in-cluster connectivity for externalTrafficPolicy=Local

borkmann opened the pull request:
See commit messages.

Fixes: #11746
Fixes: #11724

borkmann committed:
bpf: prepare in-cluster connectivity for local scope services

Currently, the in-cluster implementation for externalTrafficPolicy=Local
services is incorrect in that E-W service communication should work always
for these cases. Reason is that we can always /directly/ select a backend
from the socket lb and never need to go the additional hop where original
client source IP would be otherwise obfuscated for non-DSR. Minimal straight
forward change would be to have the orchestrator add two service entries
into the map with different scoping. By default we do the regular lookup
where we only find the set of local endpoints. This is hit from bpf_host
side as-is today. For socket lb, we then do a scope switch and do a lookup
on the cluster internal scope which can then select all backends including
remote ones. This is least-complex and also most optimial for fast-path
since for bpf_host the second lookup can simply be optimized-out from the
compiler. All the rest wrt special service type behavior is transparent.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: optimize fast-path by removing l3-lb lookup from services

It's always built-in, but there is no use-case anymore after we've removed
the stand-alone LB back in the days, and it's not configurable via services
either. Removing this logic reduces the service map lookups in critical fast
path from two to one, and allows to simplify the code where we would otherwise
need to restore the dest port in the lookup.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add scope into L3n4Addr struct

Pull in the lookup scope into L3n4Addr. This is needed since services are
hashed based on their frontend L3n4Addr. Also, implement support for setting
the scope via `cilium service update` CLI to allow for manual testing.

Follow-up patches will add surrogate internal entries from k8s side for
externalTrafficPolicy=Local to make backends cluster internal reachable.

Example from CLI config:

  # cilium service update --frontend="192.168.178.29:53" --backends="1.1.1.1:53" --id 11 --k8s-load-balancer --k8s-traffic-policy=local
  # cilium service update --frontend="192.168.178.29:53" --backends="8.8.8.8:53" --id 12 --k8s-load-balancer --k8s-traffic-policy=local --k8s-cluster-internal

  # ./cilium/cilium service list
  ID   Frontend              Service Type   Backend
  11   192.168.178.29:53     LoadBalancer   1 => 1.1.1.1:53
  12   192.168.178.29:53/i   LoadBalancer   1 => 8.8.8.8:53

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS       BACKEND ADDRESS
  192.168.178.29:53/i   0.0.0.0:0 (12) [LoadBalancer, Local]
                        8.8.8.8:53 (12)
  192.168.178.29:53     0.0.0.0:0 (11) [LoadBalancer, Local]
                        1.1.1.1:53 (11)

  * dig from outside node:

    23:27:47.341486 IP 192.168.178.28.48963 > 192.168.178.29.53: 7220+ [1au] A? heise.de. (49)
    23:27:47.341502 IP 192.168.178.29.48963 > 1.1.1.1.53: 7220+ [1au] A? heise.de. (49)
    23:27:47.350990 IP 1.1.1.1.53 > 192.168.178.29.48963: 7220 1/0/1 A 193.99.144.80 (61)
    23:27:47.350994 IP 192.168.178.29.53 > 192.168.178.28.48963: 7220 1/0/1 A 193.99.144.80 (61)

  * dig from inside node:

    23:31:00.994830 IP 192.168.178.29.47978 > 8.8.8.8.53: 27432+ [1au] A? heise.de. (49)
    23:31:01.004566 IP 8.8.8.8.53 > 192.168.178.29.47978: 27432 1/0/1 A 193.99.144.80 (53)

Restore from BPF map upon agent restart also works.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: add surrogate internally scoped entries for local traffic policy

Finally, wire-up Kubernetes side to add surrogate entries for cluster-internal
communication in case of externalTrafficPolicy=Local. We only filter local
backends for ScopeExternal. ScopeInternal will have everything.

Example from nginx deployment with externalTrafficPolicy=Local, listing from
node apoc:

  # ./cilium/cilium service list
  ID   Frontend                 Service Type   Backend
  4    10.100.125.187:80        ClusterIP      1 => 10.217.0.252:80
                                               2 => 10.217.1.86:80
  5    192.168.178.29:30465     NodePort       1 => 10.217.0.252:80
  6    0.0.0.0:30465            NodePort       1 => 10.217.0.252:80
  7    192.168.178.29:30465/i   NodePort       1 => 10.217.0.252:80
                                               2 => 10.217.1.86:80
  8    0.0.0.0:30465/i          NodePort       1 => 10.217.0.252:80
                                               2 => 10.217.1.86:80

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS          BACKEND ADDRESS
  10.100.125.187:80        0.0.0.0:0 (4) [ClusterIP]
                           10.217.0.252:80 (4)
                           10.217.1.86:80 (4)
  192.168.178.29:30465     0.0.0.0:0 (5) [NodePort, Local]
                           10.217.0.252:80 (5)
  192.168.178.29:30465/i   0.0.0.0:0 (7) [NodePort, Local]
                           10.217.0.252:80 (7)
                           10.217.1.86:80 (7)
  0.0.0.0:30465            0.0.0.0:0 (6) [NodePort, Local]
                           10.217.0.252:80 (6)
  0.0.0.0:30465/i          10.217.1.86:80 (8)
                           10.217.0.252:80 (8)
                           0.0.0.0:0 (8) [NodePort, Local]

  # kubectl get pods --all-namespaces -o wide
  NAMESPACE     NAME                           READY   STATUS    RESTARTS   AGE    IP             NODE   NOMINATED NODE   READINESS GATES
  default       nginx6-7d4b5d6bdf-7tpk4        1/1     Running   0          127m   10.217.1.86    tank   <none>           <none>
  default       nginx6-7d4b5d6bdf-wd9hf        1/1     Running   0          127m   10.217.0.252   apoc   <none>           <none>
  [...]

As can be seen the ClusterIP has all backends, the external facing NodePort
only has local backends whereas the internal facing NodePort has all backends.
The 0.0.0.0:30465 could still be optimized away since genCartesianProduct()
has no awareness from the logic in ParseService() which adds the surrogate
entries.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gandro commented on pkg/service/service.go:
```diff
@@ -237,10 +235,11 @@ func (s *Service) UpsertService(
+		// Services with trafficPolicy=Local may only use node-local backends
+		// for external scope. We implement this by filtering out all backend
+		// IPs which are not a local endpoint.
+		if onlyLocalBackends && frontend.Scope == lb.ScopeExternal &&
```
I think we should merge the `&& frontend.Scope == lb.ScopeExternal` condition into the `onlyLocalBackends` flag.

Or, alternatively, add `&& frontend.Scope == lb.ScopeExternal` check to the if-statement here:
https://github.com/cilium/cilium/blob/d34558669add991d90e2f346476b378413b9e205/pkg/service/service.go#L265-L268

Otherwise we will add the backends of the internal frontend to the `HealthCheckNodePort` server, which is only supposed to keep count of backends for the external fronted.

borkmann replied:
Good point, yes, thanks for catching!

gandro reviewed (changes requested):
Overall looks good so far! One issue with regards to `HealthCheckNodePort`

borkmann committed:
tests: enable additional externalTrafficPolicy=Local tests

... given kube-proxy-free setup supports them now, lets run these. Also
add tests from third node to make sure they fail to a node that has no
local backend.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

nebril commented on cilium/cmd/service_update.go:
```diff
@@ -28,15 +28,16 @@ import (
-	idU              uint64
-	frontend         string
-	backends         []string
+	deprecatedAddRev   bool // TODO(v1.8+): remove it
```
We didn't remove it ;_;

borkmann replied:
Yep, correct. We should remove it along with the other stuff that we deprecated.

nebril reviewed (approved):
CI and k8s bits look good to me

aanm commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -157,6 +157,11 @@ const (
 
+const (
+	ScopeExternal = 0
+	ScopeInternal = 1
```
This requires some documentation, I have no clue what this constants refers to.

aanm commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -293,19 +298,22 @@ func (l *L4Addr) DeepCopy() *L4Addr {
 type L3n4Addr struct {
 	IP net.IP
 	L4Addr
+	Scope uint8
```
Same, what is 'Scope'?

aanm commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -397,22 +418,30 @@ func (b *Backend) GetBackendModel() *models.BackendAddress {
+// String returns the L3n4Addr in the "IPv4:Port[/Scope]" format for IPv4 and
+// "[IPv6]:Port[/Scope]" format for IPv6.
 func (a *L3n4Addr) String() string {
+	scope := ""
```
nit: `s/scope := ""/var scope string/`

aanm commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -397,22 +418,30 @@ func (b *Backend) GetBackendModel() *models.BackendAddress {
+// StringWithProtocol returns the L3n4Addr in the "IPv4:Port[/Scope]/Protocol"
+// format for IPv4 and "[IPv6]:Port[/Scope]/Protocol" format for IPv6.
 func (a *L3n4Addr) StringWithProtocol() string {
+	scope := ""
```
nit: `s/scope := ""/var scope string/`

aanm commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -397,22 +418,30 @@ func (b *Backend) GetBackendModel() *models.BackendAddress {
-// StringWithProtocol returns the L3n4Addr in the "IPv4:Port/Protocol" format
-// for IPv4 and "[IPv6]:Port/Protocol" format for IPv6.
+// StringWithProtocol returns the L3n4Addr in the "IPv4:Port[/Scope]/Protocol"
+// format for IPv4 and "[IPv6]:Port[/Scope]/Protocol" format for IPv6.
```
Perhaps would be preferably to have the scope after the `/Protocol`? Usually the protocol is next to the port so it looks weird if we have something in between the two.

borkmann replied:
Agree, changed.

aanm commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -611,12 +619,21 @@ func (k *K8sWatcher) delK8sSVCs(svc k8s.ServiceID, svcInfo *k8s.Service, se *k8s
-	svcs := make([]loadbalancer.SVC, 0, len(ports))
+	if svcTrafficPolicy == loadbalancer.SVCTrafficPolicyLocal &&
+		(svcType == loadbalancer.SVCTypeLoadBalancer || svcType == loadbalancer.SVCTypeNodePort) {
+		svcSize = len(ports) * 2
```
add a comment for the why `len(ports) * 2`

aanm commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -640,12 +657,32 @@ func genCartesianProduct(
 				Type:     svcType,
 			})
+
+		if svcSize > len(ports) {
```
please add a comment to this condition, I would even create a new variable on top to be more explicit.
e.g. `isTrafficPolicyLocal := svcSize > len(ports)`

qmonnet commented on pkg/k8s/service_cache.go:
```diff
@@ -106,6 +106,8 @@ func NewServiceCache(nodeAddressing datapath.NodeAddressing) ServiceCache {
 }
 
 // GetServiceIP returns a random L3n4Addr that is backing the given Service ID.
+// The returned IP is with external scope since it's string representation might
```
```suggestion
// The returned IP is with external scope since its string representation might
```
(If you find other things to address; don't respin for that.)

borkmann replied:
Ok, I'll wait for the current CI to go through, and if all green, then fix the comment and wait only for travis.

qmonnet reviewed (approved):
Looks good, nice work!

borkmann committed:
docs: misc improvements to kube-proxy free gsg

Add an 'umbrella' section on client source IP preservation with links
to the related sections.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer reviewed (approved):
docs/cli/bpf LGTM. I didn't look so closely at services side.
